internal/core/src/index/VectorMemIndex.cpp

```diff
@@ -483,37 +483,65 @@ VectorMemIndex<T>::Build(const Config& config) {
     auto insert_files =
         GetValueFromConfig<std::vector<std::string>>(config, "insert_files");
     AssertInfo(insert_files.has_value(),
-               "insert file paths is empty when build disk ann index");
+               "insert file paths is empty when building in memory index");
     auto field_datas =
         file_manager_->CacheRawDataToMemory(insert_files.value());
 
-    int64_t total_size = 0;
-    int64_t total_num_rows = 0;
-    int64_t dim = 0;
-    for (auto data : field_datas) {
-        total_size += data->Size();
-        total_num_rows += data->get_num_rows();
-        AssertInfo(dim == 0 || dim == data->get_dim(),
-                   "inconsistent dim value between field datas!");
-        dim = data->get_dim();
-    }
-
-    auto buf = std::shared_ptr<uint8_t[]>(new uint8_t[total_size]);
-    int64_t offset = 0;
-    for (auto data : field_datas) {
-        std::memcpy(buf.get() + offset, data->Data(), data->Size());
-        offset += data->Size();
-        data.reset();
-    }
-    field_datas.clear();
-
     Config build_config;
     build_config.update(config);
     build_config.erase("insert_files");
     build_config.erase(VEC_OPT_FIELDS);
+    if (GetIndexType().find("SPARSE") == std::string::npos) {
+        int64_t total_size = 0;
+        int64_t total_num_rows = 0;
+        int64_t dim = 0;
+        for (auto data : field_datas) {
+            total_size += data->Size();
+            total_num_rows += data->get_num_rows();
+            AssertInfo(dim == 0 || dim == data->get_dim(),
+                       "inconsistent dim value between field datas!");
+            dim = data->get_dim();
+        }
 
-    auto dataset = GenDataset(total_num_rows, dim, buf.get());
-    BuildWithDataset(dataset, build_config);
+        auto buf = std::shared_ptr<uint8_t[]>(new uint8_t[total_size]);
+        int64_t offset = 0;
+        for (auto data : field_datas) {
+            std::memcpy(buf.get() + offset, data->Data(), data->Size());
```

Comment:
We can atually avoid this copy by introducing a wrapper?
This may save tons of memories.
Leave it for optimization

Reply:
that is very true. Currently we copy for both dense and sparse. I tried updating this but it requires some major refactoring to allow the field data to give up ownership of its underlying data. such a change will affect lots of code(most of them should be beneficial) and is way beyond the scope of this PR so I just keep it as is. 

I am adding a TODO.